marlowe/src/Language/Marlowe/Semantics.hs

```diff
@@ -1244,6 +1268,79 @@ instance ToJSON Contract where
       , "then" .= cont
       ]
 
+instance FromJSON TransactionInput where
+  parseJSON (Object v) =
+        TransactionInput <$> (parseSlotInterval =<< (v .: "tx_interval"))
+                         <*> ((v .: "tx_inputs") >>=
+                   withArray "Transaction input list" (\cl ->
+                     mapM parseJSON (F.toList cl)
+                                                      ))
+    where parseSlotInterval = withObject "SlotInterval" (\v ->
+            do from <- Slot <$> (withInteger =<< (v .: "from"))
+               to <- Slot <$> (withInteger =<< (v .: "to"))
+               return (from, to)
+                                                      )
+  parseJSON _ = fail "TransactionInput must be an object"
+
+instance ToJSON TransactionInput where
+  toJSON (TransactionInput (Slot from, Slot to) txInps) = object
+      [ "tx_interval" .= slotIntervalJSON
+      , "tx_inputs" .= toJSONList (map toJSON txInps)
+      ]
+    where slotIntervalJSON = object [ "from" .= from
+                                    , "to" .= to
+                                    ]
+
+instance FromJSON TransactionWarning where
+  parseJSON (String "assertion_failed") = return TransactionAssertionFailed
+  parseJSON (Object v) =
+        (TransactionNonPositiveDeposit <$> (parseJSON =<< (v .: "party"))
+                                       <*> (parseJSON =<< (v .: "in_account"))
+                                       <*> (parseJSON =<< (v .: "of_token"))
+                                       <*> (parseJSON =<< (v .: "asked_to_deposit")))
+    <|> (do maybeButOnlyPaid <- v .:? "but_only_paid"
+            case maybeButOnlyPaid :: Maybe Scientific of
+              Nothing -> TransactionNonPositivePay <$> (parseJSON =<< (v .: "account"))
+                                                   <*> (parseJSON =<< (v .: "to_payee"))
+                                                   <*> (parseJSON =<< (v .: "of_token"))
+                                                   <*> (parseJSON =<< (v .: "asked_to_pay"))
+              Just butOnlyPaid -> TransactionPartialPay <$> (parseJSON =<< (v .: "account"))
+                                                        <*> (parseJSON =<< (v .: "to_payee"))
+                                                        <*> (parseJSON =<< (v .: "of_token"))
+                                                        <*> (parseJSON =<< (v .: "asked_to_pay"))
+                                                        <*> getInteger butOnlyPaid)
+    <|> (TransactionShadowing <$> (parseJSON =<< (v .: "value_id"))
+                              <*> (parseJSON =<< (v .: "had_value"))
+                              <*> (parseJSON =<< (v .: "is_now_assigned")))
```

Comment:
again, I think `parseField` would neaten things up considerably here

Reply:
hmmm
Kris  10:06 AM
I think that parseJSON is redundant.
`.:` will do it for you implicitly.

Reply:
Oh, it seems it is true, very neat :)